File: python3/Linked Lists/remove_kth_last_node.py
from ds import ListNode

"""
Definition of ListNode:
class ListNode:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next
"""


def remove_kth_last_node(head: ListNode, k: int) -> ListNode:
    # A dummy node to ensure there's a node before 'head' in case we 
    # need to remove the head node.
    dummy = ListNode(-1)
    dummy.next = head
    trailer = leader = dummy
    # Advance 'leader' k steps ahead.  
    for _ in range(k):
        leader = leader.next
        # If k is larger than the length of the linked list, no node 
        # needs to be removed.
        if not leader:
            return head
    # Move 'leader' to the end of the linked list, keeping 'trailer'
    # k nodes behind.
    while leader.next:
        leader = leader.next
        trailer = trailer.next
    # Remove the kth node from the end.
    trailer.next = trailer.next.next
    return dummy.next

File: Java/src/LinkedLists/RemoveKthLastNode.java
package LinkedLists;

import DS.ListNode;

/*
    // Definition of ListNode:
    public class ListNode {
        public int val;
        public ListNode next;
        public ListNode(int val) { this.val = val; }
        public ListNode(int val, ListNode next) {
            this(val);
            this.next = next;
        }
    }
 */


public class RemoveKthLastNode {
    public static ListNode solution(ListNode head, int k) {
        // A dummy node to ensure there's a node before 'head' in case we
        // need to remove the head node.
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode trailer = dummy;
        ListNode leader = head;
        // Advance 'leader' k steps ahead.
        for (int i = 0; i < k; i++) {
            leader = leader.next;
            // If k is larger than the length of the linked list, no node
            // needs to be removed.
            if (leader == null) {
                return head;
            }
        }
        // Move 'leader' to the end of the linked list, keeping 'trailer'
        // k nodes behind.
        while (leader.next != null) {
            leader = leader.next;
            trailer = trailer.next;
        }
        // Remove the kth node from the end.
        trailer.next = trailer.next.next;
        return dummy.next;
    }
}
